### PR TITLE
Generic write via gRPC

### PR DESCRIPTION
**Pull request by juliusv**
This takes @brian-brazil's latest gRPC generic write implementation and does a couple of cleanups / fixes on it:

- fold metric name into labels
- die on error when constructing the gRPC client connection
- remove example binary from git
- other cleanups

To add in the future: tests, compression.

### CODE REVIEW

**Comment by tomwilkie** [comment]
👍 thanks @brian-brazil and @juliusv!

Could consider deduping samples for the same timeseries in the `Store` method, but with a small enough push timeout (currently 5s) this is unlikely to happen anyway.

**Comment by brian-brazil** [comment]
We still need to add snappy or similar to this.

**Comment by juliusv** [comment]
@brian-brazil We can add compression later, right? As far as I can see from http://www.grpc.io/docs/guides/wire.html, whether or not compression is used is indicated by a `Compressed-Flag` per message, as well as the encoding type in a `Message-Encoding` header, so that would be possible to add later without breaking something.

**Comment by brian-brazil** [comment]
From a cost perspective, compression isn't optional. It's only 20 lines of code or so.

**Comment by juliusv** [comment]
I can try adding it. I guess it would mean using https://godoc.org/google.golang.org/grpc#Compressor and https://godoc.org/google.golang.org/grpc#Decompressor with a custom compressor/decompressor that supports snappy. Or even the included gzip one if that's not worrying you too much on the CPU front?

**Comment by brian-brazil** [comment]
Gzip is quite expensive CPU wise from my testing, to the extent that if we offered a flag for it users may turn it off - which would be to the detriment of the other end's network costs (and the other end needs to pay CPU too).

Snappy was basically free CPU wise, and provided as good compression.

**Comment by juliusv** [comment]
Will add.

**Comment by juliusv** [comment]
@brian-brazil I added optional Snappy and GZip support.

What surprised me is that a given gRPC server (at least in Go) apparently only supports registering one decompressor.

**Comment by juliusv** [comment]
Meh. The CI fails. Not sure why neither the amd64-specific assembler codec nor the generic "other" one builds, but that leads to an undefined function. Investigating...

**Comment by juliusv** [comment]
Ah, hadn't checked in the updated version of snappy. Should be hopefully passing soon now...

**Comment by juliusv** [comment]
Looks all green from my side now - will merge soon if everyone's happy?

File: storage/remote/generic/generic.go [hunk 1]
```diff
@@ -62,16 +69,15 @@ func (c *Client) Store(samples model.Samples) error {
 		}
 		req.Timeseries = append(req.Timeseries, ts)
 	}
-	client := NewGenericWriteClient(c.conn)
 	ctxt, _ := context.WithTimeout(context.Background(), c.timeout)
```
Comment (fabxc) [inline storage/remote/generic/generic.go]: The cancel should be instantiated and called after we are done.

Comment (fabxc) [inline storage/remote/generic/generic.go]: The client should probably be created with a base context from which the request contexts are derived from.

File: storage/remote/generic/generic.go [hunk 2]
```diff
@@ -22,37 +23,43 @@ import (
...
 
-// Store sends a batch of samples to the http endpoint.
+// Store sends a batch of samples to the HTTP endpoint.
 func (c *Client) Store(samples model.Samples) error {
-	req := &GenericWriteRequest{}
+	req := &GenericWriteRequest{
+		Timeseries: make([]*TimeSeries, 0, len(samples)),
+	}
 	for _, s := range samples {
-		ts := &TimeSeries{
```
Comment (fabxc) [inline storage/remote/generic/generic.go]: Seems like a place where we should allocate `Labels` with the correct length.
On average this append would probably result in 3 reallocations.

File: storage/remote/generic/generic.proto [hunk 3]
```diff
@@ -15,9 +15,8 @@ syntax = "proto3";
 
 package generic;
 
-
 message Sample {
-  double value = 1;
```
Comment (fabxc) [inline storage/remote/generic/generic.proto]: Padding the `=` doesn't seem to be proto formatting standard. I don't think there's one golden one, but two-space indents and no padding is what I see in most official things.

File: storage/remote/generic/generic.proto [hunk 4]
```diff
@@ -27,10 +26,9 @@ message LabelPair {
 }
 
 message TimeSeries {
-  string name = 1;
-  repeated LabelPair labels = 2;
```
Comment (fabxc) [inline storage/remote/generic/generic.proto]: Do we want to define an ordering on labels as part of the protocol?

File: storage/remote/generic/generic.proto [hunk 5]
```diff
@@ -15,9 +15,8 @@ syntax = "proto3";
 
 package generic;
 
-
 message Sample {
-  double value = 1;
+  double value       = 1;
   int64 timestamp_ms = 2;
```
Comment (fabxc) [inline storage/remote/generic/generic.proto]: Naming seems somewhat awkward to me. What would be the overhead of having a custom timestamp message type on which we can define meaningful functions (like getting a `time.Time` etc.)?

**Comment by fabxc** [comment]
Just wondering about the `generic` naming in general. If we have no intend to include anything else as remote write path, why drag this, well.... generic prefix everywhere. Doesn't seem to add any value in terms of clarity.

File: storage/remote/generic/generic.go [hunk 6]
```diff
@@ -30,9 +30,20 @@ type Client struct {
...
+	var cp grpc.Compressor
+	switch compression {
+	case "gzip":
+		cp = grpc.NewGZIPCompressor()
+	case "snappy":
+		cp = &snappyCompressor{}
+	case "":
+		// No compression.
+	default:
+		log.Fatalf("Invalid generic remote compression type %q", compression)
```
Comment (fabxc) [inline storage/remote/generic/generic.go]: We are still using global loggers in most places but I think we shouldn't use `Fatal*`.
This constructor should return errors.

**Comment by beorn7** [comment]
I have to secand @fabxc . I don't think the API is generic at all. It is actually very _specific_ (to the internal storage model of Prometheus). Now we shouldn't call it literally "specific" everywhere, but "generic" feels like a misnomer.

**Comment by fabxc** [comment]
I would just call it our "write protocol".

I'm not 100% sure I caught everything correctly from our Saturday discussion. So the entire aspect of groups of time series belonging together and their annotation data (type/help so far) is not part of remote storage? Will it forever be dropped or would there be an out-of-band API to maintain an index of that?
Not all these points seemed entirely resolved to me.

As there should be read path, which should to the largest extent feel like it was actually designed together with the write path, I'd find it extremely helpful to get a mockup of that in the near future. Of course it doesn't have to be this PR but I'd like to avoid having two APIs that feel awkward together.
One of the benefits of REST is that the user has a very clear idea what he gets out after putting X in. Of course this isn't REST, but you get the usability idea.

**Comment by juliusv** [comment]
Ok, I'll try to come up with a better naming for the protobufs and types.

We pretty much defined the read protocol. It's going to look similar to what we already have in Frankenstein (https://github.com/tomwilkie/prometheus/blob/frankenstein/storage/remote/generic/generic.proto#L53-L61), except that you'll be able to send multiple selectors in one query for efficiency reasons, and also send along an opaque hint (either as k/v string map or one big string). I do indeed plan on also adding that to Prometheus proper soon.

File: storage/remote/generic/generic.go (hunk 1 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: Ok, seems like I can use https://godoc.org/google.golang.org/grpc#DialContext for that, although it's labeled as anexperimental API. The docstring of that function is still a bit confusing to me: "Once this function returns, the cancellation and expiration of ctx will be noop.". But the context is also attached to the returned `clientConn` object, so it does have relevance even after `DialContext` returns.

Just so I understand you correctly: So I pass a base context into there, then create a cancellable context based on that for each `Write()`? What do you mean with calling cancel "after we are done"? After the `Write()` completes (why, since it's already done then?)?

File: storage/remote/generic/generic.go (hunk 2 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: For sure, will be there in the next fixup push.

**Comment by fabxc** [comment]
Thanks, looking forward to seeing it all wired together. What about the annotation data stuff?
And I might have misunderstood at some point, but I thought there was at least some initial concern around breaking up the metric families?

I believe that this works just fine, but its the second stage where we are discarding information at an API boundary that will most likely never change. (First one is having users instrument in terms of namespaces and subsystems.)
Just want to make us aware of that once more before merging.

File: storage/remote/generic/generic.proto (hunk 3 above)
Comment (juliusv) [inline storage/remote/generic/generic.proto]: I did that to stay consistent with our existing `metrics.proto` formatting and the discussion around that on https://github.com/prometheus/client_model/pull/14/files. Don't really care about it much either way, but then we should change it everywhere.

File: storage/remote/generic/generic.proto (hunk 4 above)
Comment (juliusv) [inline storage/remote/generic/generic.proto]: Don't think that's necessary for anyone? But we can still add it later if some use case needs it - if we do it now and guarantee it, then we can't remove it again.

File: storage/remote/generic/generic.proto (hunk 5 above)
Comment (juliusv) [inline storage/remote/generic/generic.proto]: It's the same naming as in https://github.com/prometheus/client_model/blob/master/metrics.proto#L73, probably best to stay consistent?

File: storage/remote/generic/generic.go (hunk 1 above)
Comment (fabxc) [inline storage/remote/generic/generic.go]: It's just how it's supposed to be used. Might not be necessary here (no idea, tbh) but its how context receivers expect things to be done if the caller behaves according to the documentation: https://godoc.org/context#WithCancel

```
Canceling this context releases resources associated with it, so code should call cancel as soon as the operations running in this Context complete.
```

I think it's very easy to forget an the call might have spawned a bunch of goroutines, which might be dangling without the context being canceled. Better to make this a best practice everywhere.

File: storage/remote/generic/generic.proto (hunk 5 above)
Comment (fabxc) [inline storage/remote/generic/generic.proto]: Yea, right

File: storage/remote/generic/generic.proto (hunk 3 above)
Comment (fabxc) [inline storage/remote/generic/generic.proto]: Okay, was only cared about consistency as well. Wasn't considering our other proto.

**Comment by juliusv** [comment]
Yes, that was the conclusion on Saturday. We initially wanted to reuse the (extended) exposure format with HELP/TYPE annotations and all, but that turned out to be too problematic for a variety of reasons (see the "metric (with metadata etc.) vs. raw time series
" section in https://docs.google.com/document/d/1Qj_m-lbatySAU0vsVfAPG7mduN1-nevrpD85uQhi4sY/edit?usp=sharing). So in the end we all agreed to go this route and use the flat sample / time series format.

Not sure about ever supporting HELP/TYPE in this interface, but conceivably that could still be added in some other way if ever needed in the remote path.

File: storage/remote/generic/generic.go (hunk 1 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: Ok cool, good to know. I don't have a lot of experience with context best practices yet. So I'll add that cancel.

Comment (fabxc) [inline storage/remote/generic/generic.go]: Mh, the dial context is indeed confusing. 

Also I was phrasing things wrong/awkward previously. I'd just skip dialing with context now, but calling `Store()` should not instantiate a new background context. The API methods should rather get a context from the parent, i.e. `Store(ctx, samples)`.

My understanding is that there should be a logical context chain through your program. Otherwise they are really nothing but simple timeouts. It's not always convenient of course, especially when just introducing them. There's even a special TODO context for that (https://godoc.org/context#TODO).
But the caller of `Store` should ultimately be allowed to stop cancel any in-flight requests via the context.

File: storage/remote/generic/generic.go [hunk 7]
```diff
@@ -22,37 +23,43 @@ import (
...
 	return &Client{
-		conn:    conn,
 		timeout: timeout,
+		client:  NewGenericWriteClient(conn),
 	}
 }
 
-// Store sends a batch of samples to the http endpoint.
+// Store sends a batch of samples to the HTTP endpoint.
 func (c *Client) Store(samples model.Samples) error {
```
Comment (fabxc) [inline storage/remote/generic/generic.go]: Sorry for nitting all over the place after you were basically done. But wondering about the "Store" naming – we are just writing stuff somewhere and what happens to it is not necessarily storing, c.f. the example app just printing to stdout.

Probably fine for now but we should work on our naming ultimately.

File: storage/remote/generic/generic.go (hunk 1 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: Yeah, there's a lot of interesting questions about contexts - where to begin and terminate them, and whether it's worth threading them through everything or just parts of the call chain... also, on what level to apply timeouts: e.g. here in the caller of `Store()` or in `Store()` itself.

For now I think I'll create a TODO context directly in the generic `Write()` implemenation and set the timeout on that, since it's a timeout that's specifically configured for that implementation (maybe makes sense to change it once we really remove the others).

File: storage/remote/generic/generic.go (hunk 6 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: Agreed - I followed the existing scheme that we had for the other cases here, but it's probably best to propagate errors all the way back to main and then fatal there. Then we can do the same with the InfluxDB client (before removing it), where we also currently have a fatal log: https://github.com/prometheus/prometheus/blob/master/storage/remote/influxdb/client.go#L39

File: storage/remote/generic/generic.go (hunk 7 above)
Comment (juliusv) [inline storage/remote/generic/generic.go]: Yeah, we should call this `Send()` or something like that ultimately. For now, it's part of `storage/remote` and the interface for all the implementations defines the `Store()` method. I propose updating that all in a next step.

**Comment by juliusv** [comment]
So for the overall naming, how about I move the generic interface just up one directory/package level higher into the existing `remote` package and we just call it the "remote interface"? Like `remote.proto`, `RemoteClient`, etc.? The read path will also go into this later, that's why I'm not just calling it the "write interface". So this would be a first step towards preparing for the removal of all the other mechanisms. But I would still keep the overall `remote.go` and `queue_manager.go` structure and interfaces defined therein for now in this PR.

**Comment by juliusv** [comment]
(well, s/`RemoteClient`/`Client`/ because it'll already be part of the `remote` package then)

**Comment by juliusv** [comment]
@fabxc Pushed the context handling and pre-allocation fixups. Once we agree on the general naming / packaging, I'll wrap that up too.

**Comment by fabxc** [comment]
Naming SGTM. Yea, no need to rewrite the Queue Manager stuff now or anything.

File: storage/remote/generic/generic.go [hunk 8]
```diff
@@ -30,9 +30,20 @@ type Client struct {
...
-	conn, err := grpc.Dial(address, grpc.WithInsecure(), grpc.WithTimeout(timeout))
+func NewClient(address string, compression string, timeout time.Duration) *Client {
+	var cp grpc.Compressor
+	switch compression {
+	case "gzip":
+		cp = grpc.NewGZIPCompressor()
+	case "snappy":
+		cp = &snappyCompressor{}
+	case "":
+		// No compression.
```
Comment (brian-brazil) [inline storage/remote/generic/generic.go]: This should be hardcoded to snappy, with no option for gzip.

Comment (juliusv) [inline storage/remote/generic/generic.go]: It's funny the go-grpc only includes a built-in handler for gzip then and not the other compression types. I wonder why...

Comment (brian-brazil) [inline storage/remote/generic/generic.go]: Probably trying to keep the interface small.

**Comment by juliusv** [comment]
Pushed three more changes:

- removed the "generic" naming as far as possible for now
- hard-coded snappy compression
- return generic write init errors back to main

Especially take a look at the namings in `remote.proto`, because those will be part of the externally exposed interface.

**Comment by fabxc** [comment]
👍

**Comment by beorn7** [comment]
> Yes, that was the conclusion on Saturday. We initially wanted to reuse the (extended) exposure format with HELP/TYPE annotations and all, but that turned out to be too problematic for a variety of reasons 

Perhaps here is the right place then to chime in (instead of that mail I announced).

I have checked the code, and I actually don't think that using the normal exposition format would be too problematic. We ran into the relabeling issue during our discussion, and I promised to check that out. Doing the relabeling on the level of protobufs is perfectly feasible, and with proto3, it is even slightly less ugly than anticipated. Where it would get crazy, though, is where the relabeling touches pseudo time series and pseudo labels, e.g. buckets and `_count` and `_sum` and such. Those don't really exist in the data model of the exposition format. They only exist in the internal storage model of the Prometheus server, where everything has to be flattened into series of scalar values with labels. This representation leaks into public visibility via PromQL, as PromQL operates on that representation. I always considered this a wart: Users had to deal with weird `le` labels. We give them the quantiles of a summary in the same metric family although they are not aggregatable in a meaningful way (against our own best practices for metric families). We cannot tell the user about type and help and anything. 

Since relabeling acts on the flattened representation, like PromQL, it is another leak, and that one also affects the remote storage, as we obviously want to save the metrics after relabeling. Instead of containing the leaks and reducing the visibility of the internal storage model, it is spreading, thereby preventing advanced features — and I'm not worried so much anymore about the metadata, which can be solved: either it is really just metadata, in which case it can be treated lazily and separately; or it is part of the metric identity (as it is arguably the case for the type or the unit), which could essentially be modeled as labels. I am much more worried by now about the advantages of 1st class complex metric types that we are effectively throwing away forever. Quirks in the expression language would be nice to eliminate, but much more important would be the distributed dynamic histograms, or just more efficient storage of conventional histogram buckets (we wouldn't really need full indexing of each bucket, and very fine granular buckets could be saved quite efficiently, so that histograms with hundreds of buckets would be feasible). Check out the frequent occurrences of "too late" in the notes of our Saturday meeting. We kind of sleepwalked into this situation. (I shouted a lot, but nobody really seemed to be interested in waking up.)

The next thing is that both the remote storage approaches so far (Vulcan and Frankenstein) are super tightly coupled to the internal storage model of Prometheus. Even if we made the write path happen with the external exposition format, there would be very little point in doing the same for the read path (as Vulcan and Frankenstein are anyway following the internal Prometheus model, so converting back and forth would just be overhead). This is distinctly different from a "proper" generic write where arbitrary TSDBs would use Prometheus as a collector for Prometheus metrics as exposed in all their richness, ignorant of the way they are broken down or reduced internally.

In summary, it is more the harm already done in terms of proliferation of the internal storage model plus the very tight integration of the existing remote storage approaches that convinced me that we need a separate flat exchange format as in this PR. I would have much more preferred if we cleaned up the internal storage situation first before the internal model spreads everywhere, but obviously, the priorities were different for everybody, including myself. That's how life goes, I guess…
